### PR TITLE
Implement Tiered Hashing with small pool sizes, kick out nodes for correctness, talk more to fast nodes and reduce pool churn

### PR DESCRIPTION
**Pull request by aarshkshah1992**
### TODO

- [x] Run in staging  → https://github.com/ipfs/bifrost-gateway/pull/94
- [x] Unit Tests
- [x] Green CI
- [x] Investigate and fix low cache hit Saturn pcts 

**_Subsequent PRs:_**
- [ ] Traffic Mirroring to probe the unknown set instead of the current  probabilistic mechanism ? (subsequent PR)
-  [ ] Use exponential decay over the entire history of the node rather than sliding window percentiles to rank nodes

### CODE REVIEW

File: caboose.go [hunk 1]
```diff
@@ -16,6 +16,11 @@ import (
 	"github.com/ipfs/go-cid"
 )
 
+const (
+	BifrostProd    = "bifrost-prod"
+	BifrostStaging = "bifrost-staging"
+)
+
```
[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply caboose.go:23]: @willscott Yeah, had this for local testing. Will move to using env vars once we set them on Bifrost.

[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply caboose.go]: Done.

**Review by willscott — APPROVED** [review]
This seems fine for a next iteration

The "speed" calculation I think needs to look at bytes received versus wall clock time, rather than the throughput of individual requests

**Comment by aarshkshah1992** [comment]
@willscott Not sure what you mean by that speed comment. Speed is calculated as (bytes recieved)/(end - start time).

What change do we need to make here ?

**Comment by willscott** [comment]
That the speed we care about is how many bytes we got over the last minute - we'll get more accuracy by adding receiver bytes into a bucket per peer and looking at it overall, rather than sampling each individual request as individual throughput measurements

**Comment by aarshkshah1992** [comment]
@willscott But we add the speed of the individual request to a bucket for that peer and then use the P25 of that bucket to rank. Isn't that what you are saying here ?

File: fetcher.go [hunk 2]
```diff
@@ -187,7 +187,7 @@ func (p *pool) fetchResource(ctx context.Context, from string, resource string,
 				// my address
 				Range:          "",
 				Referrer:       respReq.Referer(),
-				UserAgent:      respReq.UserAgent(),
+				UserAgent:      "bifrost-" + os.Getenv(EnvironmentKey),
```
Comment (willscott) [inline fetcher.go:190]: [the embedding bifrost-gatway](https://github.com/ipfs/bifrost-gateway/blob/3c65632495c8cc6e0db55103c0d9deea918204bb/http_client.go#L17) is already setting a useful version in the user agent. we shouldn't drop that.

Comment (aarshkshah1992) [reply fetcher.go:190]: @willscott Okay, will append

File: failure_test.go [hunk 3]
```diff
@@ -260,6 +188,7 @@ func (e *ep) Setup() {
 	e.valid = true
 	e.resp = testBlock
 	e.server = httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(time.Millisecond * 20)
```
Comment (willscott) [inline failure_test.go:191]: why is this needed? / why are responses delayed?

File: fetcher.go [hunk 4]
```diff
@@ -74,14 +75,17 @@ func (p *pool) doFetch(ctx context.Context, from string, c cid.Cid, attempt int)
 }
 
 // TODO Refactor to use a metrics collector that separates the collection of metrics from the actual fetching
-func (p *pool) fetchResource(ctx context.Context, from string, resource string, mime string, attempt int, cb DataCallback) (err error) {
+// rm will be nil only for context cancellation errors
```
Comment (willscott) [inline fetcher.go:78]: this is confusing - it looks like it isn't nil even in the context cancellation case - it's just an empty ResponseMetrics objects?

File: fetcher.go [hunk 5]
```diff
@@ -232,22 +227,35 @@ func (p *pool) fetchResource(ctx context.Context, from string, resource string,
 			}
 		}
 	}
+	req.Header.Add("User-Agent", "bifrost-"+os.Getenv(EnvironmentKey))
```
Comment (willscott) [inline fetcher.go:230]: don't over-ride the provided user agent - there's already a more specific user agent set on the client that bifrost passes in and this will over-write it.

File: out.car
Comment (willscott) [inline out.car:1]: this shouldn't be committed / in history

File: metrics.go [hunk 6]
```diff
@@ -236,13 +207,34 @@ var (
...
 	CabooseMetrics.MustRegister(poolSizeMetric)
-	CabooseMetrics.MustRegister(poolHealthMetric)
 	CabooseMetrics.MustRegister(poolNewMembersMetric)
+	CabooseMetrics.MustRegister(poolRemovedFailureTotalMetric)
+	CabooseMetrics.MustRegister(poolRemovedConnFailureTotalMetric)
+	CabooseMetrics.MustRegister(poolRemovedReadFailureTotalMetric)
+	CabooseMetrics.MustRegister(poolRemovedNon2xxTotalMetric)
+	CabooseMetrics.MustRegister(poolMembersNotAddedBecauseRemovedMetric)
+	CabooseMetrics.MustRegister(poolEnoughObservationsForMainSetDurationMetric)
+	CabooseMetrics.MustRegister(poolTierChangMetric)
```
Comment (willscott) [inline metrics.go:237]: `poolTierChangeMetric`

File: pool.go [hunk 7]
```diff
@@ -6,27 +6,23 @@ import (
...
+
 	blocks "github.com/ipfs/go-block-format"
 	"github.com/ipfs/go-cid"
 	"github.com/patrickmn/go-cache"
-	"github.com/serialx/hashring"
 )
 
-const maxPoolSize = 300
-
 var (
```
Comment (willscott) [inline pool.go:23]: const

File: pool.go [hunk 8]
```diff
@@ -172,132 +87,53 @@ func (p *pool) Start() {
...
-				// It's not the best but it's a good start and we can tune as we go along.
-				if perf.latencyDigest.Count() > float64(minSuccessfulRetrievals) && perf.latencyDigest.Quantile(latencyPercentile) <= maxLatency {
-					poolWeightBumpMetric.Add(1)
-					m.weight = maxWeight
-				}
-			}
-		}
+	mt := p.th.GetPoolMetrics()
+	poolSizeMetric.WithLabelValues("unknown").Set(float64(mt.Unknown))
+	poolSizeMetric.WithLabelValues("main").Set(float64(mt.Main))
```
Comment (willscott) [inline pool.go:115]: "unknown" / "main" also as consts

File: pool.go [hunk 9]
```diff
@@ -362,9 +198,16 @@ func (p *pool) fetchBlockWith(ctx context.Context, c cid.Cid, with string) (blk
...
 
-	nodes, err := p.getNodesToFetch(cidToKey(c), with)
-	if err != nil {
-		return nil, err
+	aff := with
+	if aff == "" {
+		aff = cidToKey(c)
+	}
+
+	p.lk.Lock()
```
Comment (willscott) [inline pool.go:206]: why is this an exclusive lock?

File: pool.go [hunk 10]
```diff
@@ -509,9 +285,16 @@ func (p *pool) fetchResourceWith(ctx context.Context, path string, cb DataCallba
...
 
-	nodes, err := p.getNodesToFetch(path, with)
-	if err != nil {
-		return err
+	aff := with
+	if aff == "" {
+		aff = path
+	}
+
+	p.lk.Lock()
```
Comment (willscott) [inline pool.go:293]: likewise: why is this an exclusive lock?

File: tieredhashing/sorting.go [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+	"sort"
+
+	"github.com/asecurityteam/rolling"
+)
+
+func (t *TieredHashing) nodesSortedLatency() []nodeWithLatency {
+	var nodes []nodeWithLatency
+
+	for n, perf := range t.nodes {
+		perf := perf
```
Comment (willscott) [inline tieredhashing/sorting.go:13]: ?

File: tieredhashing/tiered_hashing.go [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+	"time"
+
+	"github.com/asecurityteam/rolling"
+
+	"github.com/patrickmn/go-cache"
+
+	"github.com/serialx/hashring"
+)
+
+// TODO Make env vars for tuning
```
Comment (willscott) [inline tieredhashing/tiered_hashing.go:15]: file an issue for this if it's not in the PR

File: failure_test.go (hunk 3 above)
Comment (aarshkshah1992) [reply failure_test.go:191]: Just want to introduce some form of delay in the tests as this is the "L1 peer that sends us blocks".

File: tieredhashing/tiered_hashing.go (hunk 12 above)
Comment (aarshkshah1992) [reply tieredhashing/tiered_hashing.go:15]: https://github.com/filecoin-saturn/caboose/pull/86

File: fetcher.go (hunk 5 above)
Comment (aarshkshah1992) [reply fetcher.go:230]: @willscott Can we append this to the existing `UserAgent` that the bifrost client has on it ? I believe L1s need this to be able to bifurcate metrics etc on their side based on environment.

File: caboose_test.go [hunk 13]
```diff
@@ -205,3 +189,79 @@ func TestResource(t *testing.T) {
 		t.Fatal("expected fall-over progress")
 	}
 }
+
+type HarnessOption func(config *caboose.Config)
+
+func WithTieredHashingOpts(opts []tieredhashing.Option) func(config *caboose.Config) {
```
[the remark this answers is not in the document]
Comment (aarshkshah1992) [reply caboose_test.go]: Done.

File: tieredhashing/tiered_hashing.go (hunk 12 above)
Comment (aarshkshah1992) [reply tieredhashing/tiered_hashing.go:15]: Done.

File: pool.go (hunk 10 above)
Comment (aarshkshah1992) [reply pool.go]: Yeah agreed this can be just a read lock. Fixed.

File: fetcher.go (hunk 4 above)
Comment (aarshkshah1992) [reply fetcher.go]: Yeah this is a remanant of a previous iteration. Fixed.

File: pool.go (hunk 8 above)
Comment (aarshkshah1992) [reply pool.go]: Done.

File: pool.go (hunk 7 above)
Comment (aarshkshah1992) [reply pool.go]: Done.

File: metrics.go (hunk 6 above)
Comment (aarshkshah1992) [reply metrics.go]: Done.

File: out.car
Comment (aarshkshah1992) [reply out.car:1]: Removed.

File: fetcher.go (hunk 5 above)
Comment (willscott) [reply fetcher.go]: appending is okay, sure

Comment (willscott) [reply fetcher.go]: or you could set environment as another header, right?

Comment (aarshkshah1992) [reply fetcher.go]: Appended to existing

File: failure_test.go (hunk 3 above)
Comment (willscott) [reply failure_test.go:191]: doing so masks race conditions / makes tests artificially slow - we ideally shouldn't need it - add TODO to further clean up tests

**Review by willscott — APPROVED** [review]
